File: C# TankAndBattery/Battery.cs
using System;
class Battery
{
    private Vector2D pos;
	private Vector2D frontLeft;
	private Vector2D frontRight;
	private Vector2D backLeft;
	private Vector2D backRight;
	private double angle;

    private const double WIDTH = 0.5;
	private const double LENGHT = 1.0;
	private const double RIGHT = WIDTH / 2;
	private const double LEFT = -WIDTH / 2;
	private const double FRONT = LENGHT;
	private const double BACK = 0;

    public Battery()
    {
        UpdatePos(pos);
    }

    public void UpdatePos( Vector2D tPos )
    {
        pos = tPos;
        frontLeft = new Vector2D(LEFT, FRONT);
        frontRight = new Vector2D(RIGHT, FRONT);
        backLeft = new Vector2D(LEFT, BACK);
        backRight = new Vector2D(RIGHT, BACK);
        Matrix33 mat = Matrix33.Initialize;
        Matrix33.Rotate(mat, (float)angle);

        frontLeft = mat * frontLeft;
        frontRight = mat * frontRight;
        backLeft = mat * backLeft;
        backRight = mat * backRight;

        Vector2D.Add(frontLeft, pos);
        Vector2D.Add(frontRight, pos);
        Vector2D.Add(backLeft, pos);
        Vector2D.Add(backRight, pos);
    }

    public void Rotate( double radian )
    {
        angle += radian;
        UpdatePos(pos);
    }

    public void Debuglog()
    {
        double bAngle = angle * 180 / 3.14;
        Console.WriteLine("砲台 位置 X:{0} Y:{1} 角度:{2}\n", pos.X, pos.Y, bAngle);
        Console.WriteLine("　　 前左 X:{0} Y:{1} \n　　 前右 X:{2} Y:{3}\n", frontLeft.X, frontLeft.Y,frontRight.X, frontRight.Y);
        Console.WriteLine("　　 後左 X:{0} Y:{1} \n　　 後ろ右 X:{2} Y:{3}\n", backLeft.X, backLeft.Y, backRight.X, backRight.Y);
    }
}

File: TankAndBattery/Battery.cpp
#include "Battery.h"

Battery::Battery(double width, double lenght)
{
	WIDTH = width;
	LENGHT = lenght;
	RIGHT = WIDTH / 2;
	LEFT = -WIDTH / 2;
	FRONT = LENGHT / 2;
	BACK = -LENGHT / 2;
	angle = 0;
	UpdatePos(Vector2D(0,0));
}

Battery::Battery(const Vector2D& tPos)
{
	UpdatePos( tPos );
}

Battery::~Battery()
{
}

void Battery::UpdatePos( const Vector2D& tPos )
{
	pos = tPos;
	point[0] = Vector2D(LEFT, FRONT);
	point[1] = Vector2D(RIGHT, FRONT);
	point[2] = Vector2D(RIGHT, BACK);
	point[3] = Vector2D(LEFT, BACK);

	Matrix33 mat;
	mat.Rotate(angle);
	point[0] = mat * point[0];
	point[1] = mat * point[1];
	point[2] = mat * point[2];
	point[3] = mat * point[3];

	point[0].Add(pos);
	point[1].Add(pos);
	point[2].Add(pos);
	point[3].Add(pos);
}

void Battery::Rotate( const double& radian )
{
	angle += radian;
	UpdatePos(pos);
}

void Battery::debuglog()
{
	double bAngle = angle * 180 / 3.14;
	printf("砲台 角度:%0.2f\n", bAngle);
	printf("　　 前左 X:%0.2f Y:%0.2f 前右 X:%0.2f Y:%0.2f\n", point[0], point[1]);
	printf("　　 後左 X:%0.2f Y:%0.2f 後右 X:%0.2f Y:%0.2f\n\n", point[3], point[2]);
}
